src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java

```diff
@@ -22,25 +25,56 @@ public RatpackCurrentTraceContext() {
   public TraceContext get() {
     return registrySupplier.get()
         .maybeGet(TraceContextHolder.class)
-        .map(TraceContextHolder::getContext)
+        .map(h -> h.context)
         .orElse(null);
   }
 
   @Override
-  public Scope newScope(TraceContext currentSpan) {
-    final TraceContext previous = get();
-    registrySupplier.get().add(TraceContextHolder.class, new TraceContextHolder(currentSpan));
-    return () -> registrySupplier.get().add(new TraceContextHolder(previous));
+  public Scope newScope(TraceContext current) {
+    final TraceContextHolder previous = registrySupplier.get()
+        .maybeGet(TraceContextHolder.class)
+        .orElse(TraceContextHolder.EMPTY);
+
+    if (current != null) {
+      registrySupplier.get().add(new TraceContextHolder(current));
+      MDC.put(TRACE_ID_KEY, current.traceIdString());
+    } else {
+      registrySupplier.get().add(TraceContextHolder.EMPTY);
+      MDC.remove(TRACE_ID_KEY);
+    }
+
+    return () -> {
+      registrySupplier.get().add(previous);
+      if (previous.context != null) {
+        MDC.put(TRACE_ID_KEY, previous.context.traceIdString());
+      } else {
+        MDC.remove(TRACE_ID_KEY);
+      }
+    };
   }
 
-  public static final class TraceContextHolder {
+  /**
+   * Used by TracedParallelBatch where its used to wrap a TraceContext and puts it in the
+   * registry for the forked execution.  This is marked deprecated as we prefer not to
+   * expose details of the RatpackCurrentTraceContext implementation.
+   *
+   * @param traceContext
+   * @return
+   */
```

Comment:
@llinder ok, I think I kinda see why you're pointing this out as being less than ideal. Guess we'll need to revisit `TracedParallelBatch` at some point. There's a couple of javadoc warnings at build time but I will take care of that in another commit.